# データセット作成

In [76]:
import pandas as pd

### csv読み込み

In [77]:
csv_path_fine_train = ("./datasets_csv/finetuning/train")
csv_path_fine_test = ("./datasets_csv/finetuning/test")
csv_path_pred_labeling = ("./datasets_csv/pred_labeling")

In [78]:
# ニュース記事

#df_finetuning_news = pd.read_csv(csv_path_fine + '/news/news_dataset.csv')
df_train_news = pd.read_csv(csv_path_pred_labeling + '/news/news_dataset.csv')

#print(df_finetuning_news)
print(df_train_news)


                   date                                              title  \
0   2021-01-02 08:02:00          エイベックス本社ビルが「築3年」で売却　入札額トップが落札できなかった奇々怪々\n   
1   2020-12-25 01:06:00  エイベックス(7860)、「増配」を発表し、配当利回り11.1％に！ 年間配当は1年間で2....   
2   2020-12-24 15:24:00            エイベックス、本社ビルの売却を発表　通期の純利益は150億円の黒字に転換へ\n   
3   2020-12-21 18:42:00     浜崎あゆみ、クリスマスと大みそかライブの有観客断念を発表　無観客生配信へ「残念ながら…」\n   
4   2020-11-07 17:01:00                 嵐や白石麻衣は大成功、音楽業界の「勝ち組」がごく一部だという現実\n   
5   2020-11-06 12:46:00               エイベックスが初の希望退職募集　コロナ禍でライブ・舞台関連事業に打撃\n   
6   2020-07-25 11:08:00  エンタメ業界は苦しい。でも…　「新型コロナ以前の世界が戻ってくることはない」エイベックスが見...   
7   2021-01-13 13:45:00  エイベックス、中国bilibiliとライセンス契約　J-POPのMVを提供　日本の大手レーベ...   
8   2020-08-03 06:10:00                    オンラインライブ配信はウィズコロナ時代の救世主になれるのか\n   
9   2021-01-13 11:26:00     ソニー、クリエイティブなエンターテインメント企業へ--世界4拠点からプレスカンファレンス\n   
10  2021-01-08 09:00:00                     2021年、日本アニメが世界トレンドへ飛躍する節目の年に\n   
11  2021-01-19 15:52:00                 コロナで収入源を失ったプロミュージシャン、過去作

In [79]:
# ニュースコメント

import os

csv_folder = [csv_path_fine_train , csv_path_fine_test, csv_path_pred_labeling]

file_count = 0
for p in csv_folder:
    file_count += sum((len(f) for _, _, f in os.walk(p + '/comments'))) - 1
#print(file_count)

for j, p in enumerate(csv_folder):
    cols = ['text', 'reply', 'good', 'bad']
    df_temp = pd.DataFrame(columns=cols)
    
    for i in range(file_count):
        n_file = str(i+1).zfill(3)
        file_name = "comment_dataset_" + n_file + ".csv"
        file_path = (p + "/comments/" + file_name)
        if not os.path.isfile(file_path):
            continue
        #print(str(i+1).zfill(3))#あとで消す
        df_cmt = pd.read_csv(file_path, index_col=0)
        #df_temp = pd.concat([df_temp, df_cmt], ignore_index=True)
              
        #代入
        if j <= 1:
            df_temp = pd.concat([df_temp, df_cmt], ignore_index=True)
        else:
            df_cmt.columns = ["feature", "reply", "good", "bad"]
            df_cmt[["feature", "good", "bad"]].to_csv("./datasets/pred_labeling/features_" + n_file + ".csv", index=False)
            #df_pred_comments = df_temp.copy()
        
    #代入
    if j == 0:
        df_fine_train_comments = df_temp.copy()
    elif j == 1:
        df_fine_test_comments = df_temp.copy()
        
print("finetuning train:\n", df_fine_train_comments)
print("finetuning test:\n", df_fine_test_comments)
#print("pred labeling:\n", df_pred_comments)

finetuning train:
                                                   text reply good bad
0    ハロプロの事務所も昨年末に移転しましたよね\n都内から都内だけど、複数の拠点を1拠点にまとめ...     1  278  53
1    日本の芸能プロダクションも\nアメリカのエージェントシステムに\n移行していますね！\nタレ...     0  235  45
2    エイベックスの件はともかくこれから都心の地価は下がるのではないか。これだけリモートワークが推...     3  120  17
3    その昔大学が都心から郊外や県外に脱出しました（中央大学や筑波大学など）。その結果として、人気...     1  107  20
4    河口湖は東京都心からやや遠いけれど、都心回帰で空きが増えた八王子の大学のキャンパスなんか移転...     2   70   5
..                                                 ...   ...  ...  ..
490    アニプレックス強いよね。単なる円盤屋だったのにいつの間にか市場の先頭にたっていた感じ。\n\n     0   49   5
491  イメージセンサーの分野でも、ソニーは負けてしまうのでしょうか？\n誰か詳しい人、教えて下さい...     0    1   1
492  ソニーも調子の良いうちにイメセン事業を上場もしくは売却するべき。\n今なら５兆円前後で売却可...     0    8  26
493  プレステ5でさらに勢いに乗りますね\n世界に通用する日本ブランドはSONYか任天堂だけになっ...     0   37  18
494                やはり不景気になっても強いのは子供と女性向けの商品なのだろう。\n\n     1    7  20

[495 rows x 4 columns]
finetuning test:
                                                  text reply good bad
0   イギリスから戻って来た\n布袋寅泰氏はどうするのかね

In [80]:
#ラベル
import os

csv_folder = [csv_path_fine_train , csv_path_fine_test]

#file_count = 0
#for p in csv_folder:
#    file_count += sum((len(f) for _, _, f in os.walk(p + '/comments'))) - 1
#print(file_count)

for j, p in enumerate(csv_folder):
    cols = ['label']
    df_temp = pd.DataFrame(columns=cols)

    for i in range(file_count):
        n_file = str(i+1).zfill(3)
        file_name = "comment_labels_" + n_file + ".csv"
        file_path = (p + "/labels/" + file_name)
        if not os.path.isfile(file_path):
            continue
        #print(str(i+1).zfill(3))#あとで消す
        df_labels = pd.read_csv(file_path)#, index_col=0
        # concat
        #print(df_labels)
        df_temp = pd.concat([df_temp, df_labels], ignore_index=True)

    #代入
    if j == 0:
        df_fine_train_labels = df_temp.copy()
    elif j == 1:
        df_fine_test_labels = df_temp.copy()
        
print("labels train:\n", df_fine_train_labels)
print("labels test:\n", df_fine_test_labels)

labels train:
         label
0    positive
1    positive
2    positive
3    negative
4    positive
..        ...
490  positive
491  negative
492  negative
493  positive
494  positive

[495 rows x 1 columns]
labels test:
        label
0   negative
1   negative
2   negative
3   negative
4   positive
5   negative
6   negative
7   negative
8   negative
9   positive
10  negative
11  negative
12  negative
13  negative
14  positive
15  negative
16  negative
17  negative
18  negative
19  positive


In [81]:
# データセットの作成と保存

datasets_folder =  ("./datasets/finetuning")


#print("finetuning train:\n", df_fine_train_comments)
#print("finetuning test:\n", df_fine_test_comments)
#print("pred labeling:\n", df_pred_comments)

df_fine_train_comments.columns = ["feature", "reply", "good", "bad"]
df_fine_train_comments["feature"].to_csv(datasets_folder + "/train/features.csv", index=False)

df_fine_test_comments.columns = ["feature", "reply", "good", "bad"]
df_fine_test_comments["feature"].to_csv(datasets_folder + "/test/features.csv", index=False)

#print("labels train:\n", df_fine_train_labels)
#print("labels test:\n", df_fine_test_labels)

df_fine_train_labels["label"].to_csv(datasets_folder + "/train/labels.csv", index=False)
df_fine_test_labels["label"].to_csv(datasets_folder + "/test/labels.csv", index=False)

# 上に持っていく
#df_pred_comments.columns = ["feature", "reply", "good", "bad"]
#df_pred_comments["feature"].to_csv("./datasets/pred_labeling/features.csv", index=False)

    

In [86]:
# 単語分割する関数を定義
import MeCab
import re
import string

def preprocessing_text(text):
    '''
    前処理
    '''
    # 改行コードを消去
    text = re.sub('<br />', '', text)
    text = re.sub('\n', '', text)

    # カンマ、ピリオド以外の記号をスペースに置換
    for p in string.punctuation:
        if (p == ".") or (p == ","):
            continue
        else:
            text = text.replace(p, " ")
            
    for p in string.punctuation:
        if (p == "。") or (p == "、"):
            continue
        else:
            text = text.replace(p, "　")
            
    # ピリオドなどの前後にはスペースを入れておく
    text = text.replace(".", " . ")
    text = text.replace(",", " , ")
    text = text.replace("。", " 。 ")
    text = text.replace("、", " 、 ")
    
    return text

#m_t = MeCab.Tagger('-Owakati -d /usr/lib/mecab/dic/mecab-ipadic-neologd')
#m_t = MeCab.Tagger('-Ochasen -d /usr/local/lib/mecab/dic/mecab-ipadic-neologd')
wakati = MeCab.Tagger("-Owakati -d /usr/local/lib/mecab/dic/mecab-ipadic-neologd")

def tokenizer_mecab(text):
    '''
    分かち書き
    '''
    words = wakati.parse(text).split()
    
    return words


In [87]:
text = '機械学習が好きです。\n＋＋？？？：：機械学習が好きです。'

pre_text = preprocessing_text(text)
tokenized_text = tokenizer_mecab(pre_text)
print("tokenized:\n", tokenized_text)

tokenized:
 ['機械学習', 'が', '好き', 'です', '。', '＋', '＋', '？？？', '：', '：', '機械学習', 'が', '好き', 'です', '。']


In [88]:
text = '機械学習が好きです。\n＋＋？？？：：機械学習が好きです。'

pre_text = preprocessing_text(text)
print("pre:\n", pre_text)
tokenized_text = tokenizer_mecab(pre_text)
print("tokenized:\n", tokenized_text)
print("text tokenized:\n", len(tokenized_text))


pre:
 機械学習が好きです 。 ＋＋？？？：：機械学習が好きです 。 
tokenized:
 ['機械学習', 'が', '好き', 'です', '。', '＋', '＋', '？？？', '：', '：', '機械学習', 'が', '好き', 'です', '。']
text tokenized:
 15


In [97]:
#!pip install bert-tensorflow
#!pip install keras-bert

In [98]:
#!pip install sentencepiece

https://tech-blog.cloud-config.jp/2020-02-06-category-classification-using-bert/

In [99]:
#!git clone --recurse-submodules https://github.com/yoheikikuta/bert-japanese

## 設定ファイル

In [100]:
import pandas as pd
import sentencepiece as spm

# 学習データX
# feature.csvは上記で用意したファイルのパスを指定してください
train_features_df = pd.read_csv('./datasets/finetuning/train/features.csv')
print("train_features_df:\n", train_features_df)


# テキストの最初に[CLS]、最後に'[SEP]をつけて単語数を数える関数
def _get_indice(feature):
    tokens = []
    tokens.append('[CLS]')
    pre_text = preprocessing_text(feature)#追加
    tokenized_text = tokenizer_mecab(pre_text)#追加
    tokens.extend(tokenized_text)#追加
    #tokens.extend(sp.encode_as_pieces(feature))# sentence piece
    tokens.append('[SEP]')
    print("テキストのデータ :\n",tokens)
    number = len(tokens)
    return number

# sentence pieceでは、与えた文書の中で高い頻度で現れるフレーズは、
# 多少長くても一つの単位として認識します。
# Mecabでは対応する辞書を使って文章を分割します。辞書にはneologdとかがよく使われます。
#これでも上手くいくことも多いですが、語彙数が大きくなってしまうことや、
# 分割の仕方が分割したいデータセットに適していないこともあり、問題点
#sp = spm.SentencePieceProcessor()

# ダウンロードした事前学習モデルのパスを指定してください
#sp.Load('./downloads/bert-wiki-ja/wiki-ja.model')

numbers = []
for feature in train_features_df['feature']:
    features_number = _get_indice(feature)
    numbers.append(features_number)

print(numbers)

# 最大トークン数
max_token_num = max(numbers)
print("max_token_number: " + str(max_token_num))

train_features_df:
                                                feature
0    ハロプロの事務所も昨年末に移転しましたよね\n都内から都内だけど、複数の拠点を1拠点にまとめ...
1    日本の芸能プロダクションも\nアメリカのエージェントシステムに\n移行していますね！\nタレ...
2    エイベックスの件はともかくこれから都心の地価は下がるのではないか。これだけリモートワークが推...
3    その昔大学が都心から郊外や県外に脱出しました（中央大学や筑波大学など）。その結果として、人気...
4    河口湖は東京都心からやや遠いけれど、都心回帰で空きが増えた八王子の大学のキャンパスなんか移転...
..                                                 ...
490    アニプレックス強いよね。単なる円盤屋だったのにいつの間にか市場の先頭にたっていた感じ。\n\n
491  イメージセンサーの分野でも、ソニーは負けてしまうのでしょうか？\n誰か詳しい人、教えて下さい...
492  ソニーも調子の良いうちにイメセン事業を上場もしくは売却するべき。\n今なら５兆円前後で売却可...
493  プレステ5でさらに勢いに乗りますね\n世界に通用する日本ブランドはSONYか任天堂だけになっ...
494                やはり不景気になっても強いのは子供と女性向けの商品なのだろう。\n\n

[495 rows x 1 columns]
テキストのデータ :
 ['[CLS]', 'ハロプロ', 'の', '事務所', 'も', '昨年末', 'に', '移転', 'し', 'まし', 'た', 'よ', 'ね', '都内', 'から', '都内', 'だ', 'けど', '、', '複数', 'の', '拠点', 'を', '1', '拠点', 'に', 'まとめ', 'た', 'みたい', '経営', '立て直し', 'する', '上', 'で', '、', '固定費', 'の', '中', 'で', 'も', 'まず', '着手', 'する', 'の', 'は', '家賃', 'です', 'から', 'ね', '

テキストのデータ :
 ['[CLS]', '「', 'ザ・ドリフターズ', '、', 'ビートたけし', '、', '渥美清', '、', '萩本欽一', '、', '美空ひばり', '、', '浅香光代', '、', '淡谷のり子', '、', '村田英雄', '、', '伊東四朗', '、', '水の江瀧子', '、', '沢村貞子', '、', '永六輔', 'ら', '・', '・', '・', '」', 'ビートたけし', '→', 'ツー・ビート', '萩本欽一', '→', 'コント55号', '伊東四朗', '→', 'てんぷくトリオ', 'と', '書い', 'て', '欲しかっ', 'た', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', '芸能事務所', 'だけ', 'の', '問題', 'じゃ', 'なく', '、', '事務所', 'そのもの', 'の', '需要', 'が', 'コロナ', '後', '考える', '会社', '多く', '出', 'て', 'くる', '。', 'コロナ', 'の', '感染', 'どこ', 'で', '終息', 'なるか', 'だ', 'な', '。', '1', '極', '集中', 'が', '悪', 'だ', 'わ', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', '浅草', 'に', '来', 'ない', 'と', '歴史', 'が', '変わる', '！', '葛飾郡', 'の', '歴史', '埋立地', 'は', 'ない', '時代', 'を', '話せ', '[SEP]']
テキストのデータ :
 ['[CLS]', '相撲部屋', 'も', '今', 'じゃ', '両国', '辺り', 'だけ', 'じゃ', 'なく', '西新井', '、', '松戸', 'に', 'も', 'ある', 'から', 'ね', '[SEP]']
テキストのデータ :
 ['[CLS]', 'テレビ', 'に', '出す', 'より', 'Youtube', 'を', '頑張っ', 'た', 'ほう', 'が', '初期', '投資', 'を', '安く', '抑え', 'れる', 'もの', 'な', '。', '撮影', 'も', '映画', 'も'

テキストのデータ :
 ['[CLS]', 'もう', '時代', '的', 'に', '終身雇用', 'は', '無理', 'だろ', 'う', '。', 'グローバル', '、', 'ダイバーシティ', '、', '競争力', 'も', '求め', 'られる', '時代', '。', '終身雇用', 'が', '無理', 'なら', 'もっと', '働き', '方', 'が', '色々', 'あっ', 'て', 'も', 'いい', 'と', '思う', '。', '完全', '副業', '解禁', 'や', '８', '時間', 'の', '縛りつけ', 'じゃ', 'なく', '個人', '個人', 'で', '時間', '等', 'の', '契約', 'を', '結ん', 'で', '柔軟', 'な', '働き', '方', 'が', 'できる', 'よう', 'に', 'する', 'と', 'か', '。', '終身雇用', 'が', '難しい', 'って', '言う', 'だけ', 'で', 'は', '経営者', '失格', '。', '何かしら', '新しい', '働き', '方', 'を', '提案', 'し', 'ない', 'と', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', '此処', 'も', '、', '松浦', 'さん', 'の', 'Yes', 'man', 'しか', '居', 'ない', 'の', 'が', '、', '滑落', 'の', '要因', 'の', '一つ', 'で', 'は', '。', '配信', 'LIVE', 'とか', '、', '苦しい', '中', 'でも', 'アイディア', '絞り', '出し', 'て', 'やっ', 'てる', 'とこ', 'は', 'やっ', 'て', 'ます', 'よ', 'ね', '。', '不測', 'の', '事態', 'と', 'は', '言え', '、', '先見の明', 'が', 'な', 'さ', '過ぎる', '気', 'が', 'し', 'ます', '。', '音楽', 'バブル期', 'を', '忘れ', 'られ', 'ず', 'に', 'いる', 'と', 'いう', 'か', '。', '[SEP]']
テキストのデータ :


テキストのデータ :
 ['[CLS]', 'みんな', '農業', 'に', '転身', 'し', 'たら', 'いい', 'のに', '。', '食費', 'は', '自家栽培', 'の', '野菜', 'に', '200', '坪', 'の', '古民家', 'が', '500万円', 'で', '購入可能', '、', '派遣会社', 'は', '20時間', '以上', '勤務', 'だ', 'と', '厚生年金', '制度', 'に', '加入', 'に', '社会保険', 'あり', '、', '大阪', 'から', '岡山県', 'に', '移住', 'を', '７', '年', '前', 'に', 'し', 'まし', 'た', '。', '高齢者', 'に', 'は', '暮らし', 'やすい', '。', '高額', 'な', '住宅ローン', 'の', '支払い', 'が', 'ある', 'の', 'に', 'リストラ', 'だ', 'なんて', '？', '生活水準', '下げる', 'べき', '。', '古民家', 'が', '6', 'DK', 'で', '小学校', '近い', 'のに', '家賃', '1万円', 'とか', '空き家バンク', '検索', 'し', 'たら', 'あり', 'ます', 'よ', 'ね', '退職金', 'で', 'ハウス', 'トマト', '栽培', '始め', 'た', '脱サラ', '組', 'も', '居', 'ます', '。', '農業', 'は', '休憩', '4時間', 'あり', 'ます', 'よ', '。', '工場', 'ライン', '作業', 'より', 'は', '全然', 'ラク', 'で', '人間関係', 'モメ', 'ないし', '。', '終わりの時', '間', 'も', '自分', '決めて', 'いい', 'し', '。', '農業', 'は', '自営業', '、', 'お米', 'は', 'その', '年', 'から', 'カネ', 'に', 'なり', 'ます', 'トラクター', 'が', '200万', 'とか', 'JA', 'で', '無料', '貸し出し', 'し', 'て', 'ます', 'し', '耕作放棄地', 'も', '余っ', '

 ['[CLS]', 'あの', '贅沢三昧', 'の', '社長', '？', '違っ', 'た', 'か', 'な', '？', 'いい', '時', 'って', 'いつまでも', '続か', 'ない', 'ねぇ', '。', '豆柴の大群', 'が', 'エーベックス', 'に', '入っ', 'たって', 'やたら', '喜んで', 'た', 'が', '今', 'は', '大丈夫', 'か', '？', '緊急', 'ミーティング', '始動', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', 'avex', 'は', '沢尻エリカ', 'が', '薬', 'が', '理由', 'で', '前', 'の', '会社', 'を', '解雇', 'さ', 'れ', 'た', 'に', 'も', 'かかわら', 'ず', 'それ', 'を', '知り', 'ながら', '、', '当時', 'の', '松浦', '社長', 'が', '沢尻エリカ', 'と', '契約', 'し', 'た', '。', 'まぁ', '正直', '、', 'まとも', 'な', '会社', 'で', 'は', 'ない', 'な', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', 'でも', '株価', 'は', 'ドンドン', '上がる', '。', '公的資金', 'が', '入っ', 'て', 'から', '、', '株価', 'は', '経済指標', 'で', 'は', 'なくなっ', 'た', 'なぁ', '。', 'まあ', '下がり', 'さえ', 'し', 'なけれ', 'ば', '、', '誰', 'も', '損', 'は', 'し', 'ない', 'けど', '。', '下がり', 'さえ', 'し', 'なけれ', 'ば', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', 'やっぱり', '生', 'ライブ', 'し', 'て', 'グッズ', '等', 'で', '収益', 'あげ', 'ない', 'と', '厳しい', 'と', '思い', 'ます', '。', 'CD', 'など', '売り上げ', 'も', '今', 'は', '難しい', 'よ', 'ね', '買っ', 'て', 

 ['[CLS]', '全盛期', 'の', '浜崎', 'は', '1人', 'で', '200', '億', '以上', '叩き', 'だし', 'て', 'い', 'た', 'から', 'なぁ', '。', 'コロナ', 'の', '影響', 'が', '怖い', '[SEP]']
テキストのデータ :
 ['[CLS]', 'これから', '、', '益々', '、', '景気', 'は', '、', '悪化', 'の', '一途', 'だろ', 'う', '…', 'どれ', 'だけ', 'の', '企業', 'が', '年', 'を', '越せる', '事', 'やら', '…', '[SEP]']
テキストのデータ :
 ['[CLS]', '青山', 'の', '一等地', 'の', '豪華', 'な', '自社ビル', '。', 'あれ', 'は', '必要', 'か', 'なー', '。', '心配', 'です', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', '自分', 'は', 'まだ', '安定', '業種', 'な', '方', 'で', '良かっ', 'た', 'か', 'な', '・', '・', 'まぁ', '安', '月給', 'だ', 'けど', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', '行く', '宛', 'が', 'ある', '有能', 'な', '人', 'が', '辞め', 'て', '行く', 'よ', 'ね', '。', 'エンタメ', 'なんて', '横', 'の', '繋がり', 'が', '大きい', 'し', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', 'この', '会社', '位', 'なら', '残業', 'し', 'て', 'も', '、', '残業代', 'は', '出', 'た', 'の', 'で', 'は', '？', '[SEP]']
テキストのデータ :
 ['[CLS]', 'もう', 'コロナ', '前', 'に', 'は', '戻ら', 'ない', 'から', 'な', '。', 'ライブ', 'も', '採算', 'ベース', 'で', 'やれ', 'ない', 'だろ', '。', 'ワクチン', '

 ['[CLS]', '熱', 'も', '出', 'ない', '風邪', 'に', 'ビクビク', 'し', 'て', '自粛', '自粛', 'マスコミ', '、', '馬鹿', '厚労省', 'の', 'お陰', 'で', '今年', 'の', '年末', 'は', '失業者', '続', '失', '[SEP]']
テキストのデータ :
 ['[CLS]', '薬', '中', '助け', 'て', '社員', 'を', '切る', '。', 'とんでも', 'ない', '会社', '。', '[SEP]']
テキストのデータ :
 ['[CLS]', 'あの', '超', '有名', 'な', 'エーベックス', 'が', '！', '[SEP]']
テキストのデータ :
 ['[CLS]', '勿論', 'ウラ', 'で', 'は', '金食い虫', 'と', 'なっ', 'てる', 'だけ', 'の', '奴', 'を', '辞め', 'させる', 'ため', '、', '肩たたき', 'する', 'ん', 'だろ', '？', '[SEP]']
テキストのデータ :
 ['[CLS]', 'この', '会社', 'も', 'あっち', 'へ', 'ころり', '、', 'そっち', 'へ', 'ころり', 'だ', 'もん', 'な', '[SEP]']
テキストのデータ :
 ['[CLS]', '明るい', 'ニュース', 'ない', 'ね', '[SEP]']
テキストのデータ :
 ['[CLS]', 'これ', 'は', 'もしや', '…', 'の', 'まね', 'この', '呪い', '…', '？', '[SEP]']
テキストのデータ :
 ['[CLS]', '社畜', 'から', '脱皮', 'する', '好機', 'だ', '[SEP]']
テキストのデータ :
 ['[CLS]', 'うわー', 'ﾟ', 'Д', 'ﾟ', 'イケイケ', 'の', 'エイベックス', 'で', 'すら', '[SEP]']
テキストのデータ :
 ['[CLS]', 'これ', 'でも', '自助', 'か', '。', '日本人', 'って', '鬼', 'だ', 'ねえ', '。', '鬼', '滅', 'の', '世界', 'だ'

## 学習

## BERTの設定ファイル、モデルのロード  
学習回数と事前に調べていた最大トークン数、ファイルパスを自分用に書き換えてください。以下に書き換える箇所を示します。  

- config_path：設定ファイルのパス  
- checkpoint_path：事前学習モデルのファイルパス  
    - 拡張子まで書かないでください  
- SEQ_LEN：最大トークン数  
- EPOCH：学習回数  

In [101]:
import sys
from keras_bert import load_trained_model_from_checkpoint

#sys.pathに追加（必要なのか調査が必要）
sys.path.append('modules')

#import pprint
#pprint.pprint(sys.path)

# BERTのロード
config_path = './downloads/bert-wiki-ja_config/bert_finetuning_config_v1.json'
# 拡張子まで記載しない（.ckptファイルで保存されている）
checkpoint_path = './downloads//bert-wiki-ja/model.ckpt-1400000'

# 最大のトークン数
SEQ_LEN = max_token_num#上の処理の出力
BATCH_SIZE = 16
BERT_DIM = 768
LR = 1e-4
# 学習回数
EPOCH = 20

# 学習ずみモデルでモデル構築
bert = load_trained_model_from_checkpoint(config_path, checkpoint_path, training=True,  trainable=True, seq_len=SEQ_LEN)
bert.summary()

# この後に追加する（転移学習）
# 分類問題用にモデルの再構築

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
Input-Token (InputLayer)        [(None, 224)]        0                                            
__________________________________________________________________________________________________
Input-Segment (InputLayer)      [(None, 224)]        0                                            
__________________________________________________________________________________________________
Embedding-Token (TokenEmbedding [(None, 224, 768), ( 24576000    Input-Token[0][0]                
__________________________________________________________________________________________________
Embedding-Segment (Embedding)   (None, 224, 768)     1536        Input-Segment[0][0]              
______________________________________________________________________________________________

## 学習データのロード関数
こちらの関数でもモデルの読み込みを行うので、各自ファイルパスの変更をお願いします。

- sp.load("ファイルパス")

## 文章のベクトル化
_get_indice関数では、SentencePieceとwikipediaモデルを使用し文章のベクトル化を行っています

## 学習データ読込
_load_labeldata関数は学習データを読込、_get_indice関数を用いて特徴量を抽出しています。




In [102]:
from keras.utils import np_utils

In [103]:
from keras import utils
import numpy as np

# ここでもsentence piece
sp = spm.SentencePieceProcessor()
sp.Load('./downloads/bert-wiki-ja/wiki-ja.model')

# 上に同じ名前の関数があるので注意
# 最大単語数分のID化された文を返す関数
# maxlenがなくてエラーになるので勝手に追加（maxlenは最大単語数か？）
def _get_indice(feature, maxlen):
#def _get_indice(feature):
    # インデックス ０で埋める
    indices = np.zeros((maxlen), dtype = np.int32)
    # 最初に[CLS]、最後に'[SEP]をつけてトークン作る
    tokens = []
    tokens.append('[CLS]')
    pre_text = preprocessing_text(feature)#追加
    tokenized_text = tokenizer_mecab(pre_text)#追加
    tokens.extend(tokenized_text)#追加
    #tokens.extend(sp.encode_as_pieces(feature))# sentence piece
    tokens.append('[SEP]')

    for t, token in enumerate(tokens):
        # 最大単語数までトークンの単語をindicesに入れていく
        if t >= maxlen:
            break
        try:
            indices[t] = sp.piece_to_id(token)# id化してくれる？
        except:
            logging.warn(f'{token} is unknown.')# コメントしてくれる
            indices[t] = sp.piece_to_id('<unk>')# id化してくれる？unknown
    
    # 最大単語数分のID化された文を返す
    return indices


#勝手に追加 maxlen=103  
# 引数のパスは直接書けばいらないかも
def _load_labeldata(train_dir, test_dir, maxlen):
    # pandasでcsvの学習データとテストデータを読み込む
    train_features_df = pd.read_csv(f'{train_dir}/features.csv')
    train_labels_df = pd.read_csv(f'{train_dir}/labels.csv')
    test_features_df = pd.read_csv(f'{test_dir}/features.csv')
    test_labels_df = pd.read_csv(f'{test_dir}/labels.csv')
    
    ##### ラベル側の処理 #####
    
    # ラベルのユニーク値を取り出す（ラベル数）（インデックスとラベル別別に保管）
    # ネガポジなら　ポジティブ, ネガティブ と　０、１　を入れてしまえばいいと思われる
    #{'スポーツ': 0, '携帯電話': 1},
    label2index = {k: i for i, k in enumerate(train_labels_df['label'].unique())}
    #{0: 'スポーツ', 1: '携帯電話'}
    index2label = {i: k for i, k in enumerate(train_labels_df['label'].unique())}
    #　クラス数（何種類に分類するか）ネガポジなら２
    class_count = len(label2index)
    
    # Numpyユーティリティ to_categorical(y, nb_classes=None)
    # クラスベクトル（0からnb_classesまでの整数）を categorical_crossentropyとともに用いるためのバイナリのクラス行列に変換します．
    # y: 行列に変換するクラスベクトル, nb_classes: 総クラス数
    # ↓trainのラベルを文字からインデックスを使用して変換
    train_labels = utils.np_utils.to_categorical([label2index[label] for label in train_labels_df['label']], num_classes=class_count)
    #　testのインデックスをまず作る
    test_label_indices = [label2index[label] for label in test_labels_df['label']]
    # ↓testのラベルを文字からインデックスを使用して変換
    test_labels = utils.np_utils.to_categorical(test_label_indices, num_classes=class_count)

    ##### 特徴量側の処理 #####
    
    train_features = []
    test_features = []
    for feature in train_features_df['feature']:
        # 上で作った関数 _get_indice  を使ってID化
        train_features.append(_get_indice(feature, maxlen))
    # shape(len(train_features), maxlen)のゼロの行列作成
    train_segments = np.zeros((len(train_features), maxlen), dtype = np.float32)

    for feature in test_features_df['feature']:
        # 上で作った関数 _get_indice  を使ってID化
        test_features.append(_get_indice(feature, maxlen))
    # shape(len(test_features), maxlen)のゼロの行列作成
    test_segments = np.zeros((len(test_features), maxlen), dtype = np.float32)

    print(f'Trainデータ数: {len(train_features_df)}, Testデータ数: {len(test_features_df)}, ラベル数: {class_count}')

    return {
        'class_count': class_count,
        'label2index': label2index,
        'index2label': index2label,
        'train_labels': train_labels,
        'test_labels': test_labels,
        'test_label_indices': test_label_indices,
        'train_features': np.array(train_features),
        'train_segments': np.array(train_segments),
        'test_features': np.array(test_features),
        'test_segments': np.array(test_segments),
        'input_len': maxlen
    }

In [104]:
_get_indice("制作", maxlen=SEQ_LEN)

array([  4, 828,   5,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   

In [105]:
_load_labeldata('./datasets/finetuning/train', './datasets/finetuning/test', maxlen=SEQ_LEN)

Trainデータ数: 495, Testデータ数: 20, ラベル数: 2


{'class_count': 2,
 'label2index': {'positive': 0, 'negative': 1},
 'index2label': {0: 'positive', 1: 'negative'},
 'train_labels': array([[1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [0., 1.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [0., 1.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [0., 1.],
        [0., 1.],
     

## モデル作成関数

In [106]:
# おそらく、この関数を作った理由は複数分類モデルを自由に作れるようにしたかったからだ。
# 単にネガポジにするなら関数にしないで直接書けばいい。

from keras.layers import Dense, Dropout, LSTM, Bidirectional, Flatten, GlobalMaxPooling1D
from keras_bert.layers import MaskedGlobalMaxPool1D
from keras import Input, Model

# nadam を選べば使わなくてもいい
# https://github.com/CyberZHG/keras-bert
from keras_bert import AdamWarmup, calc_train_steps

def _create_model(input_shape, class_count):
    # AdamWarmupをオプティマイザーとして使用するために必要な情報を得る関数
    # nadam を選べば使わなくてもいい
    decay_steps, warmup_steps = calc_train_steps(
        input_shape[0],
        batch_size=BATCH_SIZE,
        epochs=EPOCH,
    )

    # 学習済みモデル 「bert」 の最終出力層のoutputを取り出す
    bert_last = bert.get_layer(name='NSP-Dense').output
    x1 = bert_last
    # 最終出力層のoutputを新規作成した全結合層に入れる
    output_tensor = Dense(class_count, activation='softmax')(x1)
    
    # Trainableの場合は、Input Masked Layerが3番目の入力なりますが、
    # FineTuning時には必要無いので1, 2番目の入力だけ使用します。
    # Trainableでなければkeras-bertのModel.inputそのままで問題ありません。
    model = Model([bert.input[0], bert.input[1]], output_tensor)
    model.compile(loss='categorical_crossentropy',
                  optimizer=AdamWarmup(decay_steps=decay_steps, warmup_steps=warmup_steps, lr=LR),
                  #optimizer='nadam',
                  metrics=['mae', 'mse', 'acc'])

    return model

## 学習データのロードとモデルの準備
事前準備で作成した学習用データと学習後のモデル名および出力先を指定してください。

- trains_dir,tests_dir：学習用データのパス
- model_filename：学習後のモデル名、出力先のパス



In [107]:
#import keras.utils

In [108]:
#from keras.utils import np_utils

In [109]:
#from keras import Input, Model, utils
#import numpy as np

In [35]:
# データロードとモデルの準備
from keras.callbacks import EarlyStopping, ModelCheckpoint, TensorBoard

trains_dir = './datasets/finetuning/train'
tests_dir = './datasets/finetuning/test'

#上で作った関数
data = _load_labeldata(trains_dir, tests_dir, SEQ_LEN)

# モデルの読み込み
model_filename = './downloads/models/knbc_finetuning.model'

# 上で作った関数（関数を使わずに直接書くこともできる）
# data['train_features'].shape　は　文の数×最大単語数　＝　特徴量Xのインプットshape
# data['class_count']　は　クラスの数
model = _create_model(data['train_features'].shape, data['class_count'])

model.summary()

Trainデータ数: 61, Testデータ数: 10, ラベル数: 2
Model: "model_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
Input-Token (InputLayer)        [(None, 214)]        0                                            
__________________________________________________________________________________________________
Input-Segment (InputLayer)      [(None, 214)]        0                                            
__________________________________________________________________________________________________
Embedding-Token (TokenEmbedding [(None, 214, 768), ( 24576000    Input-Token[0][0]                
__________________________________________________________________________________________________
Embedding-Segment (Embedding)   (None, 214, 768)     1536        Input-Segment[0][0]              
_______________________________________________________

## 学習の実行
いよいよ学習の実行です。以下のプログラムを実行した際に画像のような出力が出ると思います。（tensorflowのバージョンでWarningが出ますが問題ありません）あとはお茶でも飲みながら学習経過を観察してみましょう！

ファインチューニング？

In [36]:
history = model.fit([data['train_features'], data['train_segments']],
          data['train_labels'],
          epochs = 3,
          #epochs = EPOCH,
          batch_size = BATCH_SIZE,
          validation_data=([data['test_features'], data['test_segments']], data['test_labels']),
          shuffle=False,
          verbose = 1,
          callbacks = [
              ModelCheckpoint(monitor='val_acc', mode='max', filepath=model_filename, save_best_only=True)
          ])


Epoch 1/3
4/4 [==============================] - 683s 155s/step - loss: 0.8308 - mae: 0.5402 - mse: 0.3145 - acc: 0.3660 - val_loss: 0.6466 - val_mae: 0.4584 - val_mse: 0.2277 - val_acc: 0.6000
INFO:tensorflow:Assets written to: ./downloads/models/knbc_finetuning.model/assets
Epoch 2/3
4/4 [==============================] - 435s 110s/step - loss: 0.4363 - mae: 0.3271 - mse: 0.1339 - acc: 0.8871 - val_loss: 0.6272 - val_mae: 0.3890 - val_mse: 0.2299 - val_acc: 0.6000
Epoch 3/3
4/4 [==============================] - 364s 96s/step - loss: 0.1853 - mae: 0.1532 - mse: 0.0432 - acc: 0.9705 - val_loss: 0.7420 - val_mae: 0.4041 - val_mse: 0.2704 - val_acc: 0.6000


## 推論時にAttention Weightを出力するようにモデルをロードする

In [112]:
#from keras.models import load_model

# custom_objects 
# custom_objects引数を使ってロード機構にそのカスタムレイヤーなどを渡すことができます
#model = load_model(model_filename, custom_objects=SeqSelfAttention.get_custom_objects())

#ModelクラスAPI   model = Model(inputs=a, outputs=b)  model = Model(inputs=[a1, a2], outputs=[b1, b2, b3])
# model.get_layer('attention').output 2つ目のアウトプットを指定
#model = Model(inputs=model.input, outputs=[model.output, model.get_layer('attention').output])
#model.summary()

## 予測 TEST

In [ ]:
tests_features_df = pd.read_csv('./datasets/pred_labeling/features_001.csv')
tests_features_df.loc[0]['feature']

In [ ]:
import types

In [8]:
import sys
import pandas as pd
import sentencepiece as spm
import logging
import numpy as np

from keras import utils
from keras.models import load_model
from keras.preprocessing.sequence import pad_sequences
from keras_bert import load_trained_model_from_checkpoint
from keras_bert import get_custom_objects
from sklearn.metrics import classification_report, confusion_matrix


#sys.pathに追加（必要なのか調査が必要）
sys.path.append('modules')

# SentencePieceProccerモデルの読込
spp = spm.SentencePieceProcessor()
spp.Load('./downloads/bert-wiki-ja/wiki-ja.model')

# BERTの学習したモデルの読込（ダウンロードした？勝手に保存される？）
model_filename = './downloads/models/knbc_finetuning.model'
model = load_model(model_filename, custom_objects=get_custom_objects())
#model = load_model(model_filename, custom_objects=SeqSelfAttention.get_custom_objects())
model = Model(inputs=model.input, outputs=[model.output, model.get_layer('Encoder-12-MultiHeadSelfAttention').output])
# ↑ここでmodel = Model(inputs=a, outputs=b) としてAttentionも出すようにする。


# 上のと同じのを入れると思われるため、消していいかも(ファイルを分けるなら必要)
#SEQ_LEN = 103#206
maxlen = SEQ_LEN
#
# 上にあったのと同じ？ → predictように変更
def _get_indice_pred(feature, maxlen):
    indices = np.zeros((maxlen), dtype=np.int32)

    tokens = []
    tokens.append('[CLS]')
    pre_text = preprocessing_text(feature)#追加
    tokenized_text = tokenizer_mecab(pre_text)#追加
    tokens.extend(tokenized_text)#追加
    #tokens.extend(spp.encode_as_pieces(feature))
    tokens.append('[SEP]')

    for t, token in enumerate(tokens):
        if t >= maxlen:
            break
        try:
            indices[t] = spp.piece_to_id(token)
        except:
            logging.warn('unknown')
            indices[t] = spp.piece_to_id('<unk>')
            
    return indices, tokens

tests_features_df = pd.read_csv('./datasets/pred_labeling/features_001.csv')
feature = tests_features_df.loc[0]['feature']
#feature = "昨日は携帯電話を買いに行った。"
#feature = "昨日はスポーツしに行った。"

test_features = []
indices, tokens = _get_indice_pred(feature, maxlen)
test_features.append(indices)

#勝手に追加
test_features = np.array(test_features)

test_segments = np.zeros(
    (len(test_features), maxlen), dtype=np.float32)


# model = Modelを使えば推定　predict[0][0]　２次元のリストで返せる。
predicted_test_labels = model.predict([test_features, test_segments])#.argmax(axis=1)
#predict = model.predict(test_features)


print("完了")

KeyboardInterrupt: 

In [17]:
tests_features_df.loc[0]['feature']

'イギリスから戻って来た\n布袋寅泰氏はどうするのかね\n緊急事態宣言に\n劇場とかは含まれないみたいだけど\n全国からファンを\n武道館に来させるのかな\n中止もしくは\n武道館に来なくても払い戻しします\n位の措置は取った方がいいと思う\n\n'

In [ ]:
print(tokens)

In [43]:

print("shape: ", len(predicted_test_labels))
#print("predicted_test_labels: ", predicted_test_labels)
print("predict: ", predicted_test_labels[0])
print("predict: ", predicted_test_labels[0].argmax(axis=1))
print("attention:", predicted_test_labels[1].shape)
print("attention:\n", predicted_test_labels[1])
print("attention:\n", predicted_test_labels[1].argmax(axis=1))#axis=1
print("attention:\n", predicted_test_labels[1].argmax(axis=1).shape)#axis=1

shape:  2
predict:  [[0.59652174 0.40347835]]
predict:  [0]
attention: (1, 214, 768)
attention:
 [[[-0.15614392  0.11198843 -0.7218799  ... -0.43427202 -0.15799451
    0.13034657]
  [-0.20469886  0.40216246 -0.79378    ... -0.36917427 -0.27232108
    0.30159944]
  [ 0.22955036  0.48005208 -0.7393608  ... -0.8082622   0.52140796
   -0.10452686]
  ...
  [-0.04775316  0.3499331  -0.30924207 ... -0.11876121 -0.22986516
    0.13258928]
  [-0.00219479  0.08754978 -0.21091847 ... -0.09936933 -0.21315327
    0.09863827]
  [-0.04547665  0.01744612 -0.15949327 ... -0.13492337 -0.17773315
    0.08752056]]]
attention:
 [[ 29  23  10  44  12  31   6  35  52  55  11  12  55  35  45  42   2  35
   18  56   4  36  35  10  36  54  35  54  34  34   4  34  34  10   5  57
    5  24  18  34  21  24  36  29  24  24  42  31  43  34   1  11  41  31
   52  31   2  57  31  35  58  34  24  17  11   5   3  58  39  34   2  22
   34  27  22  18  59  59  34 207  59  34   2  45  24  54   2  58  11  48
   54   2  22  

_____________
Encoder-12-MultiHeadSelfAttenti (None, 206, 768)     0           Encoder-11-FeedForward-Norm[0][0]
                                                                 Encoder-12-MultiHeadSelfAttention
__________________________________________________________________________________________________
Encoder-12-MultiHeadSelfAttenti (None, 206, 768)     1536        Encoder-12-MultiHeadSelfAttention

In [44]:
# 入力シーケンスはpad_sequenceにより、以下の様に0でpre paddingしています。
# [0 0 0 0 x1(300) x2(300) x3(300)]
# Attention Weightは入力シーケンスに対応して計算されるため、
# 入力シーケンスのpadding分シフトします。
weights = [w.max() for w in predicted_test_labels[1][0][-len(tokens):]]
df = pd.DataFrame([tokens, weights], index=['token', 'weight']).T

mean = np.asarray(weights).mean()#np.asarray　参照コピー

df['rank'] = df['weight'].rank(ascending=False)#ランキング
# wから平均を引いた値が0より大きいものだけ（偏差）
df['normalized'] = df['weight'].apply(lambda w: max(w - mean, 0))#行全体や列全体に対して、同じ操作
df['weight'] = df['weight'].astype('float32')
df['attention'] = df['normalized'] > 0
# df.style.background_gradient で色つけ
df = df.style.background_gradient(cmap='Blues', subset=['normalized'])

display(df)

## 予測 Excel出力

In [46]:
import sys
import pandas as pd
import sentencepiece as spm
import logging
import numpy as np

from keras import utils
from keras.models import load_model
from keras.preprocessing.sequence import pad_sequences
from keras_bert import load_trained_model_from_checkpoint
from keras_bert import get_custom_objects
from sklearn.metrics import classification_report, confusion_matrix

import openpyxl

#sys.pathに追加（必要なのか調査が必要）
sys.path.append('modules')

# 上にあったのと同じ？ → predict用に変更
def _get_indice_pred(feature, maxlen):
    indices = np.zeros((maxlen), dtype=np.int32)

    tokens = []
    tokens.append('[CLS]')
    pre_text = preprocessing_text(feature)#追加
    tokenized_text = tokenizer_mecab(pre_text)#追加
    tokens.extend(tokenized_text)#追加
    #tokens.extend(spp.encode_as_pieces(feature))
    tokens.append('[SEP]')

    for t, token in enumerate(tokens):
        if t >= maxlen:
            break
        try:
            indices[t] = spp.piece_to_id(token)
        except:
            logging.warn('unknown')
            indices[t] = spp.piece_to_id('<unk>')
            
    return indices, tokens


# SentencePieceProccerモデルの読込
spp = spm.SentencePieceProcessor()
spp.Load('./downloads/bert-wiki-ja/wiki-ja.model')

# BERTの学習したモデルの読込（ダウンロードした？勝手に保存される？）
model_filename = './downloads/models/knbc_finetuning.model'
model = load_model(model_filename, custom_objects=get_custom_objects())
#model = load_model(model_filename, custom_objects=SeqSelfAttention.get_custom_objects())
model = Model(inputs=model.input, outputs=[model.output, model.get_layer('Encoder-12-MultiHeadSelfAttention').output])
# ↑ここでmodel = Model(inputs=a, outputs=b) としてAttentionも出すようにする。


# 上のと同じのを入れると思われるため、消していいかも(ファイルを分けるなら必要)
#SEQ_LEN = 103#206
maxlen = SEQ_LEN

y_train = []
#for i in range(file_count):
for i in range(3):
    n_file = str(i+1).zfill(3)
    file_name = "features_" + n_file + ".csv"
    f_path = ("./datasets/pred_labeling/" + file_name)
    if not os.path.isfile(f_path):
        continue

    df_tests_features = pd.read_csv(f_path)
    print("File Name: ", file_name)
    
    #excelファイル保管用
    excel_file = './attention_excel/attention_' + n_file + '.xlsx'
    writer = pd.ExcelWriter(excel_file, engine='xlsxwriter')
        
    wb = openpyxl.Workbook()
    sheet = wb.active
    sheet.title = 'Cover'
    c1 = sheet["A1"]
    c2 = sheet["A2"]
    c1.value =  "Attention出力用ファイルです。"
    c2.value =  "詳細は次のシート以降を参照してください。"
    wb.save(excel_file)
        
        
    pred_list = []
    good_ratio_list = []
#    for j in range(len(df_tests_features)):
    for j in range(2):
        feature = df_tests_features.loc[j]['feature']

        test_features = []
        indices, tokens = _get_indice_pred(feature, maxlen)
        test_features.append(indices)

        #勝手に追加
        test_features = np.array(test_features)

        test_segments = np.zeros(
            (len(test_features), maxlen), dtype=np.float32)

        # model = Modelを使えば推定　predict[0][0]　２次元のリストで返せる。
        predicted = model.predict([test_features, test_segments])#.argmax(axis=1)
        #predict = model.predict(test_features)

        y_pred = predicted[0].argmax(axis=1)
        #print("tokens: ", tokens)
        #print("predict: ", y_pred[0])
        
        if y_pred[0] > 0.5:
            pred_list.append([1])
        else:
            pred_list.append([-1])
        
        
        # 高評価度算出
        good = df_tests_features.loc[j]['good']
        bad = df_tests_features.loc[j]['bad']
        
        if bad == 0:
            good_ratio = [0]
        else:
            good_ratio = [good/bad]
        
        good_ratio_list.append(good_ratio)
        
        
        # 入力シーケンスはpad_sequenceにより、以下の様に0でpre paddingしています。
        # [0 0 0 0 x1(300) x2(300) x3(300)] ←３００は (None, 11, 300) の
        # Attention Weightは入力シーケンスに対応して計算されるため、
        # 入力シーケンスのpadding分シフトします。
        #weights = [w.max() for w in predicted[1][0][-len(tokens):]]
        weights = [w.max() for w in predicted[1][0]]#[-len(tokens):]
        df = pd.DataFrame([tokens, weights], index=['token', 'weight']).T

        mean = np.asarray(weights).mean()#np.asarray　参照コピー

        df['rank'] = df['weight'].rank(ascending=False)#ランキング
        # wから平均を引いた値が0より大きいものだけ（偏差）
        df['normalized'] = df['weight'].apply(lambda w: max(w - mean, 0))#行全体や列全体に対して、同じ操作
        df['weight'] = df['weight'].astype('float32')
        df['attention'] = df['normalized'] > 0
        # df.style.background_gradient で色つけ
        df = df.style.background_gradient(cmap='Blues', subset=['normalized'])

        # excel に保存
        sheetname="comment" + str(j)
        with pd.ExcelWriter(excel_file, engine="openpyxl", mode="a") as writer:
            df.to_excel(writer, sheet_name=sheetname, index=False)
        
        #display(df)
    
    # 加重平均が０より大きいか
    y = np.array(pred_list)*np.array(good_ratio_list)
    
    if y.mean() > 0:
        y_train.append(1) 
    else:
        y_train.append(0) 
        
    print("pred_list: ", pred_list)
    print("good_ratio_list: ", good_ratio_list)
    print()

y_train_df = pd.DataFrame(y_train)
y_train_df.columns = ["label"]
y_train_df.to_csv("./datasets/y_train.csv")
print("y_train: ", y_train)
        

File Name:  features_001.csv
pred_list:  [[-1], [-1]]
good_ratio_list:  [[7.961538461538462], [11.466666666666667]]

File Name:  features_003.csv
pred_list:  [[-1], [-1]]
good_ratio_list:  [[6.413897280966768], [9.019292604501608]]

y_train:  [0, 0]


# ↑完成